### PR TITLE
Add redacted files query

### PR DESCRIPTION
**Pull request by MancunianSam**
There is only a single query here with an option to only filter by the
rows where the original file ID is null.

If there is a row where it is null, this means that there is a redacted
file with no corresponding original file. For the allChecksSucceeded
call, we check that there are no rows like this.

The same query is also used in the query used by the consignment export
but in this case, we send the file ID so this can be added to the
metadata export.

I've added in tests. For the file repository, I've tried to think of
every combination that would come up but I've probably missed some so
suggestions welcome.

### CODE REVIEW

**Review by vimleshtna — COMMENTED** [review]
Looks good. Just a couple of questions. Thanks.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala [hunk 1]
```diff
@@ -131,7 +131,8 @@ class FileService(fileRepository: FileRepository,
       ffidMetadataList <- ffidMetadataService.getFFIDMetadata(consignmentId)
       avList <- avMetadataService.getAntivirusMetadata(consignmentId)
       ffidStatus <- fileStatusService.getFileStatus(consignmentId)
-    } yield fileAndMetadataList.toFiles(avList, ffidMetadataList, ffidStatus).toList
+      redactedFilePairs <- fileRepository.getRedactedFilePairs(consignmentId)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala]: I've changed a lot of the way this works after we spoke before. I'm using Sangria Projectors which let us filter on the arguments passed in the query. If the originalFilePath isn't queried, we don't look for redacted files. 
If it is, we get the file metadata like we used to, then get the redacted files, then copy any original paths over to the first list. 

As part of this, I added a `Seq[UUID]` argument for file ids and if this is populated, will only query those.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/db/repository/FileRepository.scala [hunk 2]
```diff
@@ -23,7 +22,39 @@ class FileRepository(db: Database)(implicit val executionContext: ExecutionConte
...
+
+  def getRedactedFilePairs(consignmentId: UUID, onlyNullValues: Boolean = false): Future[Seq[RedactedFiles]] = {
+    val whereSuffix = if(onlyNullValues) {
+      """AND "FileId" IS NULL"""
+    } else {
+      ""
+    }
+    val idString = consignmentId.toString
+    val regexp = "(_R\\d*$)"
+    val similarTo = "%_R\\d*\\.%"
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/db/repository/FileRepository.scala]: Ah good point yeah. I've updated the `similarTo` and added another test in to check for it.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (vimleshtna) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala]: Nice, good work Sam.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/ConsignmentFields.scala [hunk 3]
```diff
@@ -150,10 +150,17 @@ object ConsignmentFields {
...
           val fileFilters = ctx.args.argOpt("fileFiltersInput")
-          DeferFiles(ctx.value.consignmentid, fileFilters)
-        }
+          projected match {
+            case v: Vector[ProjectedName] if v.exists(_.name == "originalFilePath") =>
+              for {
+                files <- ctx.ctx.fileService.getFileMetadata(ctx.value.consignmentid, fileFilters)
+                filesWithRedacted <- ctx.ctx.fileService.updateOriginalFilePath(ctx.value.consignmentid, files)
+              } yield filesWithRedacted
+            case _ => DeferFiles(ctx.value.consignmentid, fileFilters)
```
Comment (vimleshtna) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/ConsignmentFields.scala:161]: Just a suggestion, can we make it more generic? Maybe something like:

`  case class QueriedFileFields(originalFilePath: Boolean = false)`

and then set it in the `Files` field

 ```
      Field(
        "files",
        ListType(FileType),
        arguments = FileFiltersInputArg :: Nil,
        resolve = Projector((ctx, projected) => {
          val fileFilters = ctx.args.argOpt("fileFiltersInput")
          val queriedFileFields = QueriedFileFields(projected.exists(_.name == "originalFilePath"))
          DeferFiles(ctx.value.consignmentid, fileFilters, queriedFileFields)
        })
      ),
```
and now, we can use it in the `getMetadata()` method to check if the field is queried.

  ```
def getFileMetadata(consignmentId: UUID, fileFilters: Option[FileFilters] = None, queriedFileFields: QueriedFileFields): Future[List[File]] = {
    val filters = fileFilters.getOrElse(FileFilters())
    for {
      fileAndMetadataList <- fileRepository.getFiles(consignmentId, filters)
      ffidMetadataList <- ffidMetadataService.getFFIDMetadata(consignmentId)
      avList <- avMetadataService.getAntivirusMetadata(consignmentId)
      ffidStatus <- fileStatusService.getFileStatus(consignmentId)
      redactedFilePairs <- if (queriedFileFields.originalFilePath) fileRepository.getRedactedFilePairs(consignmentId) else Future(List())
    } yield fileAndMetadataList.toFiles(avList, ffidMetadataList, ffidStatus).toList
  }
```
The above code is from your previous code changes.

So, this way we can check other fields too (i.e `metadata, ffidMetadata, antivirusMetadata`). You don't have to do it now for other fields.

Sorry, you have changed a lot of code from the previous commit, now I'm asking to change it back to the way it was with a bit of enhancement.

**Review by vimleshtna — COMMENTED** [review]
Looks good, just a suggestion, feel free to ignore it, thanks.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/ConsignmentFields.scala (hunk 3 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/ConsignmentFields.scala]: That's a really good idea. I've implemented this. As you say, it wasn't a lot to change it, I just put back some of the changes from a previous commit.

**Review by vimleshtna — APPROVED** [review]
Looks good, thanks.

**Pull request merged**